alicat/util.py

```diff
@@ -0,0 +1,224 @@
+        await self._handle_connection()
+        response = await self.connection['reader'].read(length)
+        return response.decode().strip()
+
+    async def _readline(self):
+        """Read until a line terminator."""
+        await self._handle_connection()
+        response = await self.connection['reader'].readuntil(self.eol)
+        return response.decode().strip()
+
+    async def _write(self, command: str):
+        """Write a command and do not expect a response.
+
+        As industrial devices are commonly unplugged, this has been expanded to
+        handle recovering from disconnects.
+        """
+        await self._handle_connection()
+        self.connection['writer'].write(command.encode() + self.eol)
+
+    async def _handle_connection(self):
+        """Automatically maintain TCP connection."""
+        if self.open:
+            return
+        try:
+            await asyncio.wait_for(self._connect(), timeout=0.75)
+            self.reconnecting = False
+        except (asyncio.TimeoutError, OSError):
+            if not self.reconnecting:
+                logger.error(f'Connecting to {self.address} timed out.')
+            self.reconnecting = True
+
+    async def _handle_communication(self, command):
+        """Manage communication, including timeouts and logging."""
+        try:
+            await self._write(command)
+            future = self._readline()
+            result = await asyncio.wait_for(future, timeout=0.75)
+            self.timeouts = 0
+            return result
+        except (asyncio.TimeoutError, TypeError, OSError):
+            self.timeouts += 1
+            if self.timeouts == self.max_timeouts:
+                logger.error(f'Reading from {self.address} timed out '
+                             f'{self.timeouts} times.')
+                await self.close()
+            return None
+
+    async def close(self):
+        """Close the TCP connection."""
+        if self.open:
+            await self.connection['writer'].close()
+        self.open = False
+
+
+class SerialClient(Client):
+    """Client using a directly-connected RS232 serial device."""
+
+    def __init__(self, address=None, baudrate=19200, timeout=.15, bytesize=serial.EIGHTBITS,
+                 stopbits=serial.STOPBITS_ONE, parity=serial.PARITY_NONE):
+        """Initialize serial port."""
+        super().__init__(timeout)
+        self.address = address
+        assert type(self.address) == str
+        self.serial_details = {'baudrate': baudrate,
+                               'bytesize': bytesize,
+                               'stopbits': stopbits,
+                               'parity': parity,
+                               'timeout': timeout}
+        self.ser = serial.Serial(self.address, **self.serial_details)
+
+    async def _read(self, length: int):
+        """Read a fixed number of bytes from the device."""
+        return self.ser.read(length).decode()
+
+    async def _readline(self):
```

Comment:
I think the higher-level `write_and_read` will block with this serial implementation. There needs to be an async wait-until-ready part.

I don't know how best to do this with serial today.